homeassistant/components/media_player/plex.py

```diff
@@ -272,93 +646,309 @@ def _active_media_plexapi_type(self):
+                self.device.baseurl = "{}://{}:{}".format(
+                    urlparse(client_url).scheme,
+                    urlparse(server_url).hostname,
+                    str(urlparse(client_url).port))
 
     def set_volume_level(self, volume):
         """Set volume level, range 0..1."""
-        self.device.setVolume(int(volume * 100),
-                              self._active_media_plexapi_type)
+        if self.device:
+            self._local_client_control_fix()
+            self.device.setVolume(
+                int(volume * 100), self._active_media_plexapi_type)
+            self._volume_level = volume  # store since we can't retrieve
+
+    @property
+    def volume_level(self):
+        """Return the volume level of the client (0..1)."""
+        if self._is_player_active and self.device:
+            return self._volume_level
+
+    @property
+    def is_volume_muted(self):
+        """Return boolean if volume is currently muted."""
+        if self._is_player_active and self.device:
+            return self._volume_muted
+
+    def mute_volume(self, mute):
+        """Mute the volume.
+
+        Since we can't actually mute, we'll:
+        - On mute, store volume and set volume to 0
+        - On unmute, set volume to previously stored volume
+        """
+        if self.device:
+            self._volume_muted = mute
+            if mute:
+                self._previous_volume_level = self._volume_level
+                self.set_volume_level(0)
+            else:
+                self.set_volume_level(self._previous_volume_level)
 
     def media_play(self):
         """Send play command."""
-        self.device.play(self._active_media_plexapi_type)
+        if self.device:
+            self._local_client_control_fix()
+            self.device.play(self._active_media_plexapi_type)
 
     def media_pause(self):
         """Send pause command."""
-        self.device.pause(self._active_media_plexapi_type)
+        if self.device:
+            self._local_client_control_fix()
+            self.device.pause(self._active_media_plexapi_type)
 
     def media_stop(self):
         """Send stop command."""
-        self.device.stop(self._active_media_plexapi_type)
+        if self.device:
+            self._local_client_control_fix()
+            self.device.stop(self._active_media_plexapi_type)
+
+    def turn_off(self):
+        """Turn the client off."""
+        # Fake it since we can't turn the client off
+        self.media_stop()
 
     def media_next_track(self):
         """Send next track command."""
-        self.device.skipNext(self._active_media_plexapi_type)
+        if self.device:
+            self._local_client_control_fix()
+            self.device.skipNext(self._active_media_plexapi_type)
 
     def media_previous_track(self):
         """Send previous track command."""
-        self.device.skipPrevious(self._active_media_plexapi_type)
+        if self.device:
```

Comment:
`self.device` is not enough to know if it supports remote commands - needs to include `playback` in the `device.protocolCapabilities` (which still doesn't tell you if it supports volume controls.